feat: add shared lockbox spec (#7)
defi-wonderland/specs: specs/interop/shared-lockbox-upgrade.md

# Shared Lockbox - Upgrade and migration process

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
  - [Add the chain to the op-governed dependency set](#add-the-chain-to-the-op-governed-dependency-set)
  - [Migrate ETH liquidity from `OptimismPortal` to `SharedLockbox`](#migrate-eth-liquidity-from-optimismportal-to-sharedlockbox)
    - [`LiquidityMigrator`](#liquiditymigrator)
  - [`OptimismPortal` code upgrade](#optimismportal-code-upgrade)
- [Batch transaction process](#batch-transaction-process)
  - [Diagram](#diagram)
- [Future Considerations / Additional Notes](#future-considerations--additional-notes)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

Based on the assumption that a chain joining the dependency set is an irreversible process,
the on-chain chains list is simplified by assuming that joining the Shared Lockbox is
equivalent to joining the op-governed dependency set.

> The migration process consists of three main points:

Reviewer: ```suggestion
The upgrade process consists of three main points:
```

- Add the chain to the op-governed dependency set
- Move ETH liquidity from `OptimismPortal` to `SharedLockbox`
- Upgrade the code of `OptimismPortal` to include the `SharedLockbox` integration

> The migration process also requires that:

Reviewer: ```suggestion
This process also requires:
```
maybe same here?

- `SharedLockbox` is deployed
- `SuperchainConfig` is upgraded to manage the dependency set
- `SystemConfig` is upgraded to the interop contract version

### Add the chain to the op-governed dependency set

The `SuperchainConfig` contract will be responsible for storing and managing the dependency set.

> Its `addChain` function will add a chain to the dependency set and call the `SystemConfig` of each chain

Reviewer: ```suggestion
Its `addChain` function will be used to add the chain to the dependency set and call the `SystemConfig` of each chain
```

to keep them in sync.
It will also allowlist the corresponding `OptimismPortal`, enabling it to lock and unlock ETH from the `SharedLockbox`.
Once this process is complete, the system will be ready to process deposits and withdrawals.

### Migrate ETH liquidity from `OptimismPortal` to `SharedLockbox`

The ETH will be transferred from the `OptimismPortal` to the `SharedLockbox` using an intermediate contract.

> This contract functions similarly to upgrades using the `StorageSetter`, being updated immediately before to the real implementation.

Reviewer: i find this entire phrase a bit confusing and hard to follow

This contract is meant to be used as an intermediate step for the liquidity migration.
Its unique purpose is to transfer the whole ETH balance from `OptimismPortal` to `SharedLockbox`.
This approach avoids adding extra code to the `initialize` function, which could be prone to errors in future updates.

**Interface and properties**

**`migrateETH`**

Transfers the entire ETH balance from the `OptimismPortal` to the `SharedLockbox`.

- It MUST transfer the whole ETH balance to the `SharedLockbox` when called.

```solidity
function migrateETH() external;
```

### `OptimismPortal` code upgrade

The `OptimismPortal` will start locking and unlocking ETH through the `SharedLockbox`.
It will continue to handle deposits and withdrawals but won't directly hold the ETH liquidity.
To set this up, the upgrade function will be called via `ProxyAdmin` to implement the new code,
which includes the necessary `SharedLockbox` integration.
The `SharedLockbox` address will be set during the `initialize` function. After this step,
the `OptimismPortal` will not be able to process deposits and withdrawals until the chain is registered
in `SuperchainConfig`.

## Batch transaction process

> The approach consists on handling the entire migration process in a single batched transaction.
> This transaction will consist of:

Reviewer 2: ```suggestion
The approach consists of handling the entire migration process in a single batched transaction.
This transaction will include:
```

1. Call `addChain` in the `SuperchainConfig`
   - Sending chain ID + system config address
2. Call `upgradeAndCall` in the `ProxyAdmin` for the `OptimismPortal`
   - Update provisionally to the `LiquidityMigrator` to transfer the whole ETH balance to the `SharedLockbox` in this call.
3. Call `upgrade` in the `ProxyAdmin` for the `OptimismPortal`
   - The `SharedLockbox` address is set as immutable in the new implementation

The L1 ProxyAdmin owner (L1PAO) will execute this transaction. As the entity responsible for updating contracts,

> it has the authority to perform the first two steps.

Reviewer: ```suggestion
it has the authority to perform the second and third.
```

> For the third step, the L1PAO has to be set as authorized for adding a chain to the op-governed dependency set

Reviewer: ```suggestion
For the first step, the L1PAO has to be set as authorized for adding a chain to the op-governed dependency set
```

Its sole purpose is to transfer the ETH balance.
This approach eliminates the need for additional code to move the liquidity to the lockbox later.

**Set `SuperchainConfig` as dependency manager on `SystemConfig`**

The `SystemConfig` manages the dependency updates from L1 to L2 by making a deposit through the `OptimismPortal`.
It uses access control, allowing only a designated `dependencyManager` to call it.
The `SuperchainConfig` must be set as the `dependencyManager` since it will be the contract
responsible for handling the op-governed dependency set. This step is crucial before adding a
chain to the dependency set, as it is necessary for keeping the dependency set synchronized
between L1 and L2.

To accomplish this, the code needs to be updated to `SystemConfigInterop` and `initialize` be called
with the dependency manager address

**Add the chain to the op-governed dependency set**

The `SuperchainConfig` contract will be responsible for storing and managing the dependency set.
Its `addChain` function will add a chain to the dependency set and call the `SystemConfig` of each chain
to keep them in sync.

> It will also whitelist the corresponding `OptimismPortal`, enabling it to lock and unlock ETH from the `SharedLockbox`.

Reviewer 3: ```suggestion
It will also allowlist the corresponding `OptimismPortal`, enabling it to lock and unlock ETH from the `SharedLockbox`.
```

    participant SystemConfigProxy as SystemConfig
    participant StorageSetter
    participant SuperchainConfig
    participant OptimismPortalProxy as OptimismPortal
    participant LiquidityMigrator
    participant SharedLockbox

    Note over L1PAO: Start batch

    %% Step 1: Upgrade SystemConfig to StorageSetter to zero out initialized slot
    L1PAO->>ProxyAdmin: upgradeAndCall()
    ProxyAdmin->>SystemConfigProxy: Upgrade to StorageSetter
    SystemConfigProxy->>StorageSetter: Call to set initialized slot to zero

    %% Step 2: Upgrade SystemConfig and initialize with SuperchainConfig
    L1PAO->>ProxyAdmin: upgradeAndCall()
    ProxyAdmin->>SystemConfigProxy: Upgrade to new SystemConfig implementation
    ProxyAdmin->>SystemConfigProxy: Call initialize(...SuperchainConfig address)

    %% Step 3: Add chain to SuperchainConfig
    L1PAO->>SuperchainConfig: addChain(chainId, SystemConfig address)

    %% Step 4: Upgrade OptimismPortal to intermediate implementation that transfers ETH
    L1PAO->>ProxyAdmin: upgradeAndCall()
    ProxyAdmin->>OptimismPortalProxy: Upgrade to LiquidityMigrator
    OptimismPortalProxy->>LiquidityMigrator: Call migrateETH()

>     OptimismPortalProxy->>SharedLockbox: Transfer entire ETH balance

Reviewer 2: NIT: Feel free to skip it if inside the mermaid this is not very important
```suggestion
    OptimismPortalProxy->>SharedLockbox: Transfer the entire ETH balance
```

Reviewer 2: wasn't updated, did you miss it or is not necessary?

## Overview

Based on the decision that a chain joining the dependency set is an irreversible process,
the on-chain chains list is simplified by assuming that joining the Shared Lockbox is
equivalent to joining the op-governed dependency set.

The migration process consists of four main points:

- `OptimismPortal` code upgrade (includes the `SharedLockbox` integration)
- Move ETH liquidity from `OptimismPortal` to `SharedLockbox`
- Set `SuperchainConfig` as dependency manager on `SystemConfig`
- Add the chain to the op-governed dependency set

The migration process needs some prerequisites:

- `SharedLockbox` deployed
- `SuperchainConfig` upgraded to manage the dependency set

**`OptimismPortal` code upgrade**

The `OptimismPortal` will start locking and unlocking ETH through the `SharedLockbox`.
It will continue to handle deposits and withdrawals but won't directly hold the ETH liquidity.
To set this up, the upgrade function will be called via `ProxyAdmin` to implement the new code,
which includes the necessary `SharedLockbox` integration.
The `SharedLockbox` address will be set during the `initialize` function. After this step,

> the portal will not be able to process deposits and withdrawals until the chain is registered in `SuperchainConfig`.

Reviewer 2: ```suggestion
the `OptimismPortal` will not be able to process deposits and withdrawals until the chain is registered in `SuperchainConfig`.
```

**Migrate ETH liquidity from `OptimismPortal` to `SharedLockbox`**

The ETH will be transferred from the `OptimismPortal` to the `SharedLockbox` using an intermediate contract.
This contract functions similarly to the `StorageSetter`, being updated prior to the real implementation.
Its sole purpose is to transfer the ETH balance.
This approach eliminates the need for additional code to move the liquidity to the lockbox later.

**Set `SuperchainConfig` as dependency manager on `SystemConfig`**

> The `SystemConfig` manages the dependency updates from L1 to L2 by making a deposit through the portal.

Reviewer 2: ```suggestion
The `SystemConfig` manages the dependency updates from L1 to L2 by making a deposit through the `OptimismPortal`.
```

# Shared Lockbox - Upgrade and migration process

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [Batch transaction process](#batch-transaction-process)
- [Future Considerations / Additional Notes](#future-considerations--additional-notes)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

Based on the decision that a chain joining the dependency set is an irreversible process,
the on-chain chains list is simplified by assuming that joining the Shared Lockbox is
equivalent to joining the op-governed dependency set.

The migration process consists of four main points:

- `OptimismPortal` code upgrade (includes the `SharedLockbox` integration)
- Move ETH liquidity from `OptimismPortal` to `SharedLockbox`
- Set `SuperchainConfig` as dependency manager on `SystemConfig`
- Add the chain to the op-governed dependency set

The migration process needs some prerequisites:

- `SharedLockbox` deployed
- `SuperchainConfig` upgraded to manage the dependency set

> **`OptimismPortal` code upgrade**

Reviewer 3: This one feels like the text could be divided into bulletpoints the way the text is written, perhaps it adds more clarity, could be good to check

The `OptimismPortal` will start locking and unlocking ETH through the `SharedLockbox`.
It will continue to handle deposits and withdrawals but won't directly hold the ETH liquidity.
To set this up, the upgrade function will be called via `ProxyAdmin` to implement the new code,
which includes the necessary `SharedLockbox` integration.
The `SharedLockbox` address will be set during the `initialize` function. After this step,
the portal will not be able to process deposits and withdrawals until the chain is registered in `SuperchainConfig`.

**Migrate ETH liquidity from `OptimismPortal` to `SharedLockbox`**

The ETH will be transferred from the `OptimismPortal` to the `SharedLockbox` using an intermediate contract.

> This contract functions similarly to the `StorageSetter`, being updated prior to the real implementation.

Author: hmm dont know if there is any documentation on that

to keep them in sync.
It will also whitelist the corresponding `OptimismPortal`, enabling it to lock and unlock ETH from the `SharedLockbox`.
Once this process is complete, the system will be ready to process deposits and withdrawals.

## Batch transaction process

The most efficient approach is to handle the entire migration process in a single batched transaction.
This transaction will consist of:

1. Call `upgradeAndCall` in the `ProxyAdmin` for the `SystemConfig`
   - Updating provisionally to the `StorageSetter` to zero out the initialized slot.
2. Call `upgradeAndCall` in the `ProxyAdmin` for the `SystemConfig`
   - Calling `initialize` with the `SuperchainConfig` address as the dependency manager
3. Call `addChain` in the `SuperchainConfig`
   - Sending chain ID + system config address
4. Call `upgradeAndCall` in the `ProxyAdmin` for the `OptimismPortal`
   - Update provisionally to the `LiquidityMigrator` to transfer the whole ETH balance to the `SharedLockbox` in this call.
5. Call `upgrade` in the `ProxyAdmin` for the `OptimismPortal`
   - The `SharedLockbox` address is set as immutable in the new implementation

> 

Reviewer 3: Could be nice to add a mermaid here after the steps

The L1 ProxyAdmin owner will execute this transaction. As the entity responsible for updating contracts,
it has the authority to perform the first two steps.

> For the third step, the L1PAO has to be set as authorized for adding a chain to the op-governed dependency set

Reviewer 3: what is L1PAO? if it has not been defined before, use the full name, unless that is the full name
